// @backstage-community/plugin-cost-insights/dist/utils/loading.esm.js
import { DEFAULT_DURATION } from './duration.esm.js';

var DefaultLoadingAction = /* @__PURE__ */ ((DefaultLoadingAction2) => {
  DefaultLoadingAction2["UserGroups"] = "user-groups";
  DefaultLoadingAction2["LastCompleteBillingDate"] = "billing-date";
  DefaultLoadingAction2["CostInsightsInitial"] = "cost-insights-initial";
  DefaultLoadingAction2["CostInsightsPage"] = "cost-insights-page";
  DefaultLoadingAction2["CostInsightsProducts"] = "cost-insights-products";
  DefaultLoadingAction2["CostInsightsAlerts"] = "cost-insights-alerts";
  return DefaultLoadingAction2;
})(DefaultLoadingAction || {});
const INITIAL_LOADING_ACTIONS = [
  "user-groups" /* UserGroups */,
  "cost-insights-initial" /* CostInsightsInitial */,
  "cost-insights-products" /* CostInsightsProducts */
];
const getDefaultState = (loadingActions) => {
  return loadingActions.reduce(
    (defaultState, action) => ({ ...defaultState, [action]: true }),
    {}
  );
};
const getResetState = (loadingActions) => {
  return loadingActions.reduce(
    (defaultState, action) => ({ ...defaultState, [action]: false }),
    {}
  );
};
const getResetStateWithoutInitial = (loadingActions) => {
  return loadingActions.reduce((defaultState, action) => {
    const loadingActionState = INITIAL_LOADING_ACTIONS.includes(
      action
    ) ? false : true;
    return { ...defaultState, [action]: loadingActionState };
  }, {});
};
const settledResponseOf = (responses) => {
  return responses.map(
    (response) => response.status === "fulfilled" ? response.value : null
  );
};
const initialStatesOf = (products, responses) => {
  return products.map((product, index) => ({
    entity: responses[index],
    product,
    duration: DEFAULT_DURATION
  }));
};

export { DefaultLoadingAction, INITIAL_LOADING_ACTIONS, getDefaultState, getResetState, getResetStateWithoutInitial, initialStatesOf, settledResponseOf };
