// @backstage-community/plugin-cost-insights/dist/components/AlertInsights/AlertInsightsSection.esm.js
import React from 'react';
import Box from '@material-ui/core/Box';
import Button from '@material-ui/core/Button';
import SnoozeIcon from '@material-ui/icons/AccessTime';
import AcceptIcon from '@material-ui/icons/Check';
import DismissIcon from '@material-ui/icons/Delete';
import { AlertInsightsSectionHeader } from './AlertInsightsSectionHeader.esm.js';
import { isSnoozeEnabled, isAcceptEnabled, isDismissEnabled } from '../../utils/alerts.esm.js';

const AlertInsightsSection = ({
  alert,
  number,
  onSnooze,
  onAccept,
  onDismiss
}) => {
  const isSnoozeButtonDisplayed = isSnoozeEnabled(alert);
  const isAcceptButtonDisplayed = isAcceptEnabled(alert);
  const isDismissButtonDisplayed = isDismissEnabled(alert);
  const isButtonGroupDisplayed = isSnoozeButtonDisplayed || isAcceptButtonDisplayed || isDismissButtonDisplayed;
  return /* @__PURE__ */ React.createElement(Box, { display: "flex", flexDirection: "column", mb: 6 }, /* @__PURE__ */ React.createElement(AlertInsightsSectionHeader, { alert, number }), isButtonGroupDisplayed && /* @__PURE__ */ React.createElement(Box, { display: "flex", alignItems: "center", mb: 4 }, isAcceptButtonDisplayed && /* @__PURE__ */ React.createElement(Box, { mr: 1 }, /* @__PURE__ */ React.createElement(
    Button,
    {
      color: "primary",
      variant: "contained",
      "aria-label": "accept",
      onClick: () => onAccept(alert),
      startIcon: /* @__PURE__ */ React.createElement(AcceptIcon, null)
    },
    "Accept"
  )), isSnoozeButtonDisplayed && /* @__PURE__ */ React.createElement(Box, { mr: 1 }, /* @__PURE__ */ React.createElement(
    Button,
    {
      color: "default",
      variant: "outlined",
      "aria-label": "snooze",
      disableElevation: true,
      onClick: () => onSnooze(alert),
      startIcon: /* @__PURE__ */ React.createElement(SnoozeIcon, null)
    },
    "Snooze"
  )), isDismissButtonDisplayed && /* @__PURE__ */ React.createElement(
    Button,
    {
      color: "secondary",
      variant: "outlined",
      "aria-label": "dismiss",
      disableElevation: true,
      onClick: () => onDismiss(alert),
      startIcon: /* @__PURE__ */ React.createElement(DismissIcon, null)
    },
    "Dismiss"
  )), alert.element);
};

export { AlertInsightsSection };
