// @backstage-community/plugin-cost-insights/dist/components/CostInsightsLayout/CostInsightsLayout.esm.js
import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import { CostInsightsTabs } from '../CostInsightsTabs/CostInsightsTabs.esm.js';
import { Page, Header } from '@backstage/core-components';

const useStyles = makeStyles((theme) => ({
  root: {
    gridArea: "pageContent"
  },
  header: {
    boxShadow: "none"
  },
  content: {
    padding: theme.spacing(4)
  }
}));
const CostInsightsLayout = ({
  title = "Cost Insights",
  groups,
  children
}) => {
  const classes = useStyles();
  return /* @__PURE__ */ React.createElement(Page, { themeId: "tool" }, /* @__PURE__ */ React.createElement(
    Header,
    {
      style: { boxShadow: "none" },
      title: "Cost Insights",
      pageTitleOverride: title,
      type: "Tool"
    }
  ), /* @__PURE__ */ React.createElement("div", { className: classes.root }, /* @__PURE__ */ React.createElement(CostInsightsTabs, { groups }), /* @__PURE__ */ React.createElement("div", { className: classes.content }, children)));
};

export { CostInsightsLayout };
