// @backstage-community/plugin-grafana/dist/alpha/entityCards.esm.js
import React from 'react';
import { EntityCardBlueprint } from '@backstage/plugin-catalog-react/alpha';
import { isDashboardSelectorAvailable, isAlertSelectorAvailable, isOverviewDashboardAvailable } from '../constants.esm.js';

const entityGrafanaDashboardsCard = EntityCardBlueprint.make({
  name: "dashboards",
  params: {
    filter: (entity) => Boolean(isDashboardSelectorAvailable(entity)),
    loader: () => import('../components/DashboardsCard/index.esm.js').then((m) => /* @__PURE__ */ React.createElement(m.DashboardsCard, null))
  }
});
const entityGrafanaAlertsCard = EntityCardBlueprint.make({
  name: "alerts",
  params: {
    filter: (entity) => Boolean(isDashboardSelectorAvailable(entity)) || isAlertSelectorAvailable(entity),
    loader: () => import('../components/AlertsCard/index.esm.js').then((m) => /* @__PURE__ */ React.createElement(m.AlertsCard, null))
  }
});
const entityGrafanaOverviewDashboardViewer = EntityCardBlueprint.make({
  name: "overview-dashboard",
  params: {
    filter: (entity) => isOverviewDashboardAvailable(entity),
    loader: () => import('../components/DashboardViewer/index.esm.js').then((m) => /* @__PURE__ */ React.createElement(m.EntityDashboardViewer, null))
  }
});

export { entityGrafanaAlertsCard, entityGrafanaDashboardsCard, entityGrafanaOverviewDashboardViewer };
