// @backstage-community/plugin-cost-insights/dist/components/AlertInstructionsLayout/AlertInstructionsLayout.esm.js
import React from 'react';
import Box from '@material-ui/core/Box';
import Button from '@material-ui/core/Button';
import Container from '@material-ui/core/Container';
import { makeStyles } from '@material-ui/core/styles';
import ChevronLeftIcon from '@material-ui/icons/ChevronLeft';
import { CostInsightsThemeProvider } from '../CostInsightsPage/CostInsightsThemeProvider.esm.js';
import { ConfigProvider } from '../../hooks/useConfig.esm.js';
import { CurrencyProvider } from '../../hooks/useCurrency.esm.js';
import '../../hooks/useFilters.esm.js';
import '../../hooks/useGroups.esm.js';
import '../../hooks/useLoading.esm.js';
import '../../hooks/useScroll.esm.js';
import '../../hooks/useLastCompleteBillingDate.esm.js';
import { Page, Header } from '@backstage/core-components';

const useStyles = makeStyles((theme) => ({
  root: {
    gridArea: "pageContent",
    padding: theme.spacing(4)
  }
}));
const AlertInstructionsLayout = ({
  title,
  children
}) => {
  const classes = useStyles();
  return /* @__PURE__ */ React.createElement(CostInsightsThemeProvider, null, /* @__PURE__ */ React.createElement(ConfigProvider, null, /* @__PURE__ */ React.createElement(CurrencyProvider, null, /* @__PURE__ */ React.createElement(Page, { themeId: "tool" }, /* @__PURE__ */ React.createElement(
    Header,
    {
      title: "Cost Insights",
      pageTitleOverride: title,
      type: "Tool"
    }
  ), /* @__PURE__ */ React.createElement(Container, { maxWidth: "md", disableGutters: true, className: classes.root }, /* @__PURE__ */ React.createElement(Box, { mb: 3 }, /* @__PURE__ */ React.createElement(
    Button,
    {
      variant: "outlined",
      startIcon: /* @__PURE__ */ React.createElement(ChevronLeftIcon, null),
      href: "/cost-insights"
    },
    "Back to Cost Insights"
  )), children)))));
};

export { AlertInstructionsLayout };
