// @roadiehq/backstage-plugin-argo-cd/dist/components/DetailsDrawer.esm.js
import { makeStyles, Button, Drawer, Grid, Typography, IconButton } from '@material-ui/core';
import React from 'react';
import { StructuredMetadataTable } from '@backstage/core-components';
import OpenInNewIcon from '@material-ui/icons/OpenInNew';
import CloseIcon from '@material-ui/icons/Close';
import MaterialButton from '@material-ui/core/Button';
import { useApi, configApiRef } from '@backstage/core-plugin-api';

const useStyles = makeStyles({
  paper: {
    padding: "2em"
  },
  header: {
    display: "flex",
    flexDirection: "row",
    paddingBottom: "2em"
  },
  button: {
    textTransform: "none",
    justifyContent: "flex-start",
    fontWeight: "bold"
  },
  icon: {
    fontSize: 20
  },
  content: {
    height: "80%"
  }
});
const DetailsDrawerComponent = (rowData, baseUrl) => {
  const classes = useStyles();
  const [state, setState] = React.useState(false);
  const configApi = useApi(configApiRef);
  const namespaced = configApi.getOptionalBoolean("argocd.namespacedApps") ?? false;
  const toggleDrawer = (open) => (event) => {
    if (event.type === "keydown" && (event.key === "Tab" || event.key === "Shift")) {
      return;
    }
    setState(open);
  };
  const tableContent = {
    "Argo CD Instance": rowData.metadata?.instance?.name ?? "",
    repoUrl: rowData.spec?.source?.repoURL,
    repoPath: rowData.spec?.source?.path,
    destinationServer: rowData.spec?.destination?.server,
    destinationNamespace: rowData.spec?.destination?.namespace,
    syncStatus: rowData.status?.sync?.status,
    images: rowData.status?.summary?.images,
    ...baseUrl && {
      link: /* @__PURE__ */ React.createElement(
        MaterialButton,
        {
          variant: "outlined",
          color: "primary",
          size: "small",
          title: "Open Argo CD Dashboard",
          endIcon: /* @__PURE__ */ React.createElement(OpenInNewIcon, null),
          target: "_blank",
          href: `${baseUrl}/applications/${namespaced ? `${rowData.metadata.namespace}/${rowData.metadata.name}` : rowData.metadata.name}`
        },
        "Open Argo CD Dashboard"
      )
    }
  };
  const drawerContents = () => /* @__PURE__ */ React.createElement(React.Fragment, null, /* @__PURE__ */ React.createElement("div", { className: classes.header }, /* @__PURE__ */ React.createElement(
    Grid,
    {
      container: true,
      direction: "column",
      justifyContent: "flex-start",
      alignItems: "flex-start"
    },
    /* @__PURE__ */ React.createElement(Grid, { item: true }, /* @__PURE__ */ React.createElement(Typography, { variant: "h5" }, "Argo Name: ", rowData.metadata.name))
  ), /* @__PURE__ */ React.createElement(
    IconButton,
    {
      key: "dismiss",
      title: "Close the drawer",
      onClick: () => setState(false),
      color: "inherit"
    },
    /* @__PURE__ */ React.createElement(CloseIcon, null)
  )), /* @__PURE__ */ React.createElement("div", { className: classes.content }, /* @__PURE__ */ React.createElement(StructuredMetadataTable, { metadata: tableContent })));
  return /* @__PURE__ */ React.createElement(React.Fragment, null, /* @__PURE__ */ React.createElement(
    Button,
    {
      title: rowData.metadata.name,
      className: classes.button,
      onClick: toggleDrawer(true)
    },
    rowData.metadata.name
  ), /* @__PURE__ */ React.createElement(
    Drawer,
    {
      anchor: "right",
      classes: { paper: classes.paper },
      open: state,
      onClose: toggleDrawer(false)
    },
    drawerContents()
  ));
};

export { DetailsDrawerComponent };
