// @backstage-community/plugin-gcp-projects/dist/package.json.esm.js
var name = "@backstage-community/plugin-gcp-projects";
var version = "0.8.0";
var description = "A Backstage plugin that helps you manage projects in GCP";
var backstage = {
	role: "frontend-plugin",
	pluginId: "gcp-projects",
	pluginPackages: [
		"@backstage-community/plugin-gcp-projects"
	]
};
var publishConfig = {
	access: "public",
	main: "dist/index.esm.js",
	types: "dist/index.d.ts"
};
var keywords = [
	"backstage",
	"google cloud"
];
var homepage = "https://backstage.io";
var repository = {
	type: "git",
	url: "https://github.com/backstage/community-plugins",
	directory: "workspaces/gcp-projects/plugins/gcp-projects"
};
var license = "Apache-2.0";
var sideEffects = false;
var main = "src/index.ts";
var types = "src/index.ts";
var files = [
	"dist"
];
var scripts = {
	build: "backstage-cli package build",
	clean: "backstage-cli package clean",
	lint: "backstage-cli package lint",
	prepack: "backstage-cli package prepack",
	postpack: "backstage-cli package postpack",
	start: "backstage-cli package start",
	test: "backstage-cli package test"
};
var dependencies = {
	"@backstage/core-components": "^0.17.1",
	"@backstage/core-plugin-api": "^1.10.6",
	"@material-ui/core": "^4.12.2",
	"@react-hookz/web": "^24.0.0",
	"@types/react": "^16.13.1 || ^17.0.0"
};
var devDependencies = {
	"@backstage/cli": "^0.32.0",
	"@backstage/dev-utils": "^1.1.9",
	"@testing-library/dom": "^10.0.0",
	"@testing-library/jest-dom": "^6.0.0",
	"@testing-library/react": "^15.0.0",
	"@types/react-dom": "^18.2.19",
	react: "^16.13.1 || ^17.0.0 || ^18.0.0",
	"react-dom": "^16.13.1 || ^17.0.0 || ^18.0.0",
	"react-router-dom": "6.0.0-beta.0 || ^6.3.0"
};
var peerDependencies = {
	react: "^16.13.1 || ^17.0.0 || ^18.0.0",
	"react-dom": "^16.13.1 || ^17.0.0 || ^18.0.0",
	"react-router-dom": "6.0.0-beta.0 || ^6.3.0"
};
var packageinfo = {
	name: name,
	version: version,
	description: description,
	backstage: backstage,
	publishConfig: publishConfig,
	keywords: keywords,
	homepage: homepage,
	repository: repository,
	license: license,
	sideEffects: sideEffects,
	main: main,
	types: types,
	files: files,
	scripts: scripts,
	dependencies: dependencies,
	devDependencies: devDependencies,
	peerDependencies: peerDependencies
};

export { backstage, packageinfo as default, dependencies, description, devDependencies, files, homepage, keywords, license, main, name, peerDependencies, publishConfig, repository, scripts, sideEffects, types, version };
