// @backstage/plugin-catalog-backend-module-github/dist/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var githubCatalogModule = require('./module/githubCatalogModule.cjs.js');
var GithubLocationAnalyzer = require('./analyzers/GithubLocationAnalyzer.cjs.js');
var GithubDiscoveryProcessor = require('./processors/GithubDiscoveryProcessor.cjs.js');
var GithubMultiOrgReaderProcessor = require('./processors/GithubMultiOrgReaderProcessor.cjs.js');
var GithubOrgReaderProcessor = require('./processors/GithubOrgReaderProcessor.cjs.js');
var GithubEntityProvider = require('./providers/GithubEntityProvider.cjs.js');
var GithubMultiOrgEntityProvider = require('./providers/GithubMultiOrgEntityProvider.cjs.js');
var GithubOrgEntityProvider = require('./providers/GithubOrgEntityProvider.cjs.js');
var defaultTransformers = require('./lib/defaultTransformers.cjs.js');
require('@backstage/catalog-model');
require('lodash');
require('@octokit/core');
require('@octokit/plugin-throttling');
var deprecated = require('./deprecated.cjs.js');



exports.default = githubCatalogModule.githubCatalogModule;
exports.GithubLocationAnalyzer = GithubLocationAnalyzer.GithubLocationAnalyzer;
exports.GithubDiscoveryProcessor = GithubDiscoveryProcessor.GithubDiscoveryProcessor;
exports.GithubMultiOrgReaderProcessor = GithubMultiOrgReaderProcessor.GithubMultiOrgReaderProcessor;
exports.GithubOrgReaderProcessor = GithubOrgReaderProcessor.GithubOrgReaderProcessor;
exports.GithubEntityProvider = GithubEntityProvider.GithubEntityProvider;
exports.GithubMultiOrgEntityProvider = GithubMultiOrgEntityProvider.GithubMultiOrgEntityProvider;
exports.GithubOrgEntityProvider = GithubOrgEntityProvider.GithubOrgEntityProvider;
exports.defaultOrganizationTeamTransformer = defaultTransformers.defaultOrganizationTeamTransformer;
exports.defaultUserTransformer = defaultTransformers.defaultUserTransformer;
exports.GitHubEntityProvider = deprecated.GitHubEntityProvider;
exports.GitHubOrgEntityProvider = deprecated.GitHubOrgEntityProvider;
