// @backstage-community/plugin-github-actions/dist/alpha/entityCards.esm.js
import { jsx } from 'react/jsx-runtime';
import { EntityCardBlueprint } from '@backstage/plugin-catalog-react/alpha';
import { isGithubActionsAvailable } from '../components/Router.esm.js';

const entityGithubActionsCard = EntityCardBlueprint.make({
  name: "workflow-runs",
  params: {
    filter: isGithubActionsAvailable,
    loader: () => import('../components/Router.esm.js').then((m) => /* @__PURE__ */ jsx(m.Router, { view: "cards" }))
  }
});
const entityLatestGithubActionRunCard = EntityCardBlueprint.makeWithOverrides({
  name: "latest-workflow-run",
  config: {
    schema: {
      props: (z) => z.object({
        branch: z.string().default("master")
      }).default({})
    }
  },
  factory(originalFactory, { config }) {
    return originalFactory({
      filter: isGithubActionsAvailable,
      loader: async () => import('../components/Cards/index.esm.js').then((m) => /* @__PURE__ */ jsx(m.LatestWorkflowRunCard, { ...config.props }))
    });
  }
});
const entityLatestGithubActionsForBranchCard = EntityCardBlueprint.makeWithOverrides({
  name: "latest-branch-workflow-runs",
  config: {
    schema: {
      props: (z) => z.object({
        branch: z.string().default("master")
      }).default({})
    }
  },
  factory(originalFactory, { config }) {
    return originalFactory({
      filter: isGithubActionsAvailable,
      loader: async () => import('../components/Cards/index.esm.js').then((m) => /* @__PURE__ */ jsx(m.LatestWorkflowsForBranchCard, { ...config.props }))
    });
  }
});
const entityRecentGithubActionsRunsCard = EntityCardBlueprint.makeWithOverrides({
  name: "recent-workflow-runs",
  config: {
    schema: {
      props: (z) => z.object({
        branch: z.string().default("master"),
        dense: z.boolean().default(false),
        limit: z.number().default(5).optional()
      }).default({})
    }
  },
  factory(originalFactory, { config }) {
    return originalFactory({
      filter: isGithubActionsAvailable,
      loader: async () => import('../components/Cards/index.esm.js').then((m) => /* @__PURE__ */ jsx(m.RecentWorkflowRunsCard, { ...config.props }))
    });
  }
});

export { entityGithubActionsCard, entityLatestGithubActionRunCard, entityLatestGithubActionsForBranchCard, entityRecentGithubActionsRunsCard };
