// @backstage-community/plugin-cost-insights/dist/components/ActionItems/ActionItemCard.esm.js
import React from 'react';
import classnames from 'classnames';
import Card from '@material-ui/core/Card';
import CardHeader from '@material-ui/core/CardHeader';
import '../../hooks/useConfig.esm.js';
import '../../hooks/useCurrency.esm.js';
import '../../hooks/useFilters.esm.js';
import '../../hooks/useGroups.esm.js';
import '../../hooks/useLoading.esm.js';
import { useScroll } from '../../hooks/useScroll.esm.js';
import '../../hooks/useLastCompleteBillingDate.esm.js';
import { useActionItemCardStyles } from '../../utils/styles.esm.js';

const ActionItemCard = ({
  alert,
  avatar,
  number,
  disableScroll = false
}) => {
  const classes = useActionItemCardStyles();
  const rootClasses = classnames(classes.root, {
    [classes.activeRoot]: !disableScroll
  });
  const [, setScroll] = useScroll();
  const onActionItemClick = () => {
    if (!disableScroll && number) {
      setScroll(`alert-${number}`);
    }
  };
  return /* @__PURE__ */ React.createElement(Card, { className: classes.card, raised: false, onClick: onActionItemClick }, /* @__PURE__ */ React.createElement(
    CardHeader,
    {
      classes: {
        root: rootClasses,
        action: classes.action,
        title: classes.title
      },
      title: alert.title,
      subheader: alert.subtitle,
      avatar
    }
  ));
};

export { ActionItemCard };
